farm/modeling/optimization.py

```diff
@@ -1,461 +1,136 @@
-        # warning for exceeding t_total (only active with warmup_linear
-        if (
-            not nowarn
-            and self.warn_t_total
-            and progress > 1.0
-            and progress > self.warned_for_t_total_at_progress
-        ):
-            logger.warning(
-                "Training beyond specified 't_total'. Learning rate multiplier set to {}. Please set 't_total' of {} correctly.".format(
-                    ret, self.__class__.__name__
-                )
-            )
-            self.warned_for_t_total_at_progress = progress
-        # end warning
-        return ret
-
-    @abc.abstractmethod
-    def get_lr_(self, progress):
-        """
-        :param progress:    value between 0 and 1 (unless going beyond t_total steps) specifying training progress
-        :return:            learning rate multiplier for current update
-        """
-        return 1.0
-
-
-class ConstantLR(_LRSchedule):
-    def get_lr_(self, progress):
-        return 1.0
-
-
-class WarmupCosineSchedule(_LRSchedule):
-    """
-    Linearly increases learning rate from 0 to 1 over `warmup` fraction of training steps.
-    Decreases learning rate from 1. to 0. over remaining `1 - warmup` steps following a cosine curve.
-    If `cycles` (default=0.5) is different from default, learning rate follows cosine function after warmup.
-    """
-
-    warn_t_total = True
-
-    def __init__(self, warmup=0.002, t_total=-1, cycles=0.5, **kw):
-        """
-        :param warmup:      see LRSchedule
-        :param t_total:     see LRSchedule
-        :param cycles:      number of cycles. Default: 0.5, corresponding to cosine decay from 1. at progress==warmup and 0 at progress==1.
-        :param kw:
-        """
-        super(WarmupCosineSchedule, self).__init__(warmup=warmup, t_total=t_total, **kw)
-        self.cycles = cycles
+try:
+    from apex import amp
 
-    def get_lr_(self, progress):
-        if progress < self.warmup:
-            return progress / self.warmup
-        else:
-            progress = (progress - self.warmup) / (
-                1 - self.warmup
-            )  # progress after warmup
-            return 0.5 * (1.0 + math.cos(math.pi * self.cycles * 2 * progress))
+    AMP_AVAILABLE = True
+except ImportError:
+    AMP_AVAILABLE = False
 
+logger = logging.getLogger(__name__)
 
-class WarmupCosineWithHardRestartsSchedule(WarmupCosineSchedule):
-    """
-    Linearly increases learning rate from 0 to 1 over `warmup` fraction of training steps.
-    If `cycles` (default=1.) is different from default, learning rate follows `cycles` times a cosine decaying
-    learning rate (with hard restarts).
-    """
 
-    def __init__(self, warmup=0.002, t_total=-1, cycles=1.0, **kw):
-        super(WarmupCosineWithHardRestartsSchedule, self).__init__(
-            warmup=warmup, t_total=t_total, cycles=cycles, **kw
-        )
-        assert cycles >= 1.0
+def initialize_optimizer(model,
+                         optim_opts,
```

Comment:
As mentioned, it would be great if we can hide some of the complexity for unexperienced users by providing decent defaults. Repeating `optim_opts={'name': 'AdamW', 'lr': 2E-05}` in every example, tutorial, test etc. is also not super nice. 

What do you think about some signature like this: 
``` 
initialize_optimizer(model,
                         n_batches,
                         n_epochs,
                         lr,
                         optimizer_name="AdamW",
                         optimizer_opts=None,
                         schedule_name="WarmupLinearSchedule",
                         schedule_opts=None,
                         warmup_proportion=0.1,
                         use_amp=None,
                         grad_acc_steps=1,
                         local_rank=-1,
                         log_learning_rate=False):
``` 
I am also happy to tackle that afterwards, if you don't have time to work on this.